File: Sprint7/H/C++/main.cpp
// https://contest.yandex.ru/contest/25596/problems/H/
#include <iostream>
#include <vector>
#include <algorithm>

int main() {
    int n, m;
    std::cin >> n >> m;

    std::vector<std::vector<int>> field(n, std::vector<int>(m));
    std::vector<std::vector<int>> dp(n, std::vector<int>(m, 0));

    for (int i = 0; i < n; ++i) {
        for (int j = 0; j < m; ++j) {
            char ch;
            std::cin >> ch;
            field[i][j] = ch - '0';
        }
    }

    dp[n - 1][0] = field[n - 1][0];

    for (int j = 1; j < m; ++j) {
        dp[n - 1][j] = dp[n - 1][j - 1] + field[n - 1][j];
    }

    for (int i = n - 2; i >= 0; --i) {
        dp[i][0] = dp[i + 1][0] + field[i][0];
    }

    for (int i = n - 2; i >= 0; --i) {
        for (int j = 1; j < m; ++j) {
            dp[i][j] = field[i][j] + std::max(dp[i + 1][j], dp[i][j - 1]);
        }
    }

    std::cout << dp[0][m - 1] << std::endl;

    return 0;
}

File: Sprint7/H/Rust/src/main.rs
// https://contest.yandex.ru/contest/25596/problems/H/

fn main() {
    let mut lines = std::io::stdin().lines();
    let n_m = lines
        .next()
        .unwrap()
        .unwrap()
        .split_whitespace()
        .map(|x| x.parse::<usize>().unwrap())
        .collect::<Vec<usize>>();
    let n = n_m[0];
    let m = n_m[1];

    let mut field = vec![Vec::new(); n];
    let mut dp = vec![vec![0; m]; n];

    for row in field.iter_mut() {
        let line = lines.next().unwrap().unwrap();
        for ch in line.chars() {
            row.push(ch.to_digit(10).unwrap());
        }
    }

    dp[n - 1][0] = field[n - 1][0];

    for j in 1..m {
        dp[n - 1][j] = dp[n - 1][j - 1] + field[n - 1][j];
    }

    for i in (0..(n - 1)).rev() {
        dp[i][0] = dp[i + 1][0] + field[i][0];
    }

    for i in (0..(n - 1)).rev() {
        for j in 1..m {
            dp[i][j] = field[i][j] + std::cmp::max(dp[i + 1][j], dp[i][j - 1]);
        }
    }

    println!("{}", dp[0][m - 1]);
}
